File: Day272_2025/Q272_GFG.java
import java.util.*;

class Q272_GFG {
    public int maxSubarrSum(int[] arr, int a, int b) {
        int n = arr.length;
        int ans = Integer.MIN_VALUE;

        int[] pre = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            pre[i] = pre[i - 1] + arr[i - 1];
        }

        Deque<Integer> q = new ArrayDeque<>();

        for (int i = a; i <= n; i++) {
            // remove indices out of valid window
            while (!q.isEmpty() && q.peekFirst() + b < i) {
                q.pollFirst();
            }

            // maintain increasing order of prefix sums
            while (!q.isEmpty() && pre[q.peekLast()] > pre[i - a]) {
                q.pollLast();
            }

            q.addLast(i - a);

            ans = Math.max(ans, pre[i] - pre[q.peekFirst()]);
        }

        return ans;
    }
}

File: Day272_2025/Q272_GFG.py
from collections import deque

class Solution:
    def maxSubarrSum(self, arr, a, b):
        n = len(arr)
        ans = float("-inf")

        pre = [0] * (n + 1)
        for i in range(1, n + 1):
            pre[i] = pre[i - 1] + arr[i - 1]

        q = deque()

        for i in range(a, n + 1):
            # remove indices out of window
            while q and q[0] + b < i:
                q.popleft()

            # maintain increasing order of prefix sums
            while q and pre[q[-1]] > pre[i - a]:
                q.pop()

            q.append(i - a)

            ans = max(ans, pre[i] - pre[q[0]])

        return ans
